source/extensions/filters/http/cache/thundering_herd_handler.cc

```diff
@@ -0,0 +1,233 @@
+      // destruction which can trigger another thundering herd action.
+      decoder_callbacks->continueDecoding();
+    } else if (it->second.retry_until_ > time_source_.systemTime()) {
+      decoder_callbacks->dispatcher().post([weak_filter, &request_headers]() {
+        if (auto filter = weak_filter.lock()) {
+          filter->retryHeaders(request_headers);
+        }
+      });
+    } else if (it->second.blockers_.size() < parallel_requests_) {
+      ENVOY_STREAM_LOG(debug, "ThunderingHerdHandler adding blocker {}", *decoder_callbacks,
+                       request_headers.Path()->value().getStringView());
+      it->second.blockers_.push_back(maybeMakeTimeout(key, decoder_callbacks->dispatcher()));
+      lock.Release();
+      // Lock must be released before continueDecoding, because it can trigger filter
+      // destruction which can trigger another thundering herd action.
+      decoder_callbacks->continueDecoding();
+    } else {
+      ENVOY_STREAM_LOG(debug, "ThunderingHerdHandler adding blocked {}", *decoder_callbacks,
+                       request_headers.Path()->value().getStringView());
+      addBlocked(it->second, weak_filter, decoder_callbacks, request_headers);
+    }
+  }
+
+  void addBlocked(Queue& queue, std::weak_ptr<ThunderingHerdRetryInterface> weak_filter,
+                  Http::StreamDecoderFilterCallbacks* decoder_callbacks,
+                  Http::RequestHeaderMap& request_headers) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    queue.blocked_.push_back(
+        {weak_filter, decoder_callbacks, decoder_callbacks->dispatcher(), request_headers});
+  }
+
+  /**
+   * Returns nullptr if unblock_additional_request_period was not configured.
+   * Returns a timer that will expire after the configured duration if it was configured;
+   * the timer will release one item from the queue.
+   */
+  Event::TimerPtr maybeMakeTimeout(const Key& key, Event::Dispatcher& dispatcher) {
+    if (unblock_additional_request_period_ == std::chrono::milliseconds(0)) {
+      return {};
+    }
+    Event::TimerPtr timer = dispatcher.createTimer(
+        [handler = shared_from_this(), key]() { handler->selectNewBlocker(key); });
+    timer->enableTimer(unblock_additional_request_period_);
+    return timer;
+  }
+
+  std::function<void()> onUndeliverable(const Key& key) {
+    return [key, handler = shared_from_this()]() {
+      handler->handleInsertFinished(key, InsertResult::Failed);
+    };
+  }
+
+  void selectNewBlocker(const Key& key, Queue& queue) ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_) {
+    if (queue.blocked_.empty()) {
+      ENVOY_LOG(debug, "ThunderingHerdHandler ended blocker for {}", key.path());
+      if (queue.blockers_.empty()) {
+        ENVOY_LOG(debug, "ThunderingHerdHandler deleted queue for {}", key.path());
+        queues_.erase(key);
+      }
+      return;
+    }
+    Queue::Blocked unblocking = std::move(queue.blocked_.front());
```

Comment:
In case some sort of overload is causing requests to time out, I suggest unblocking the most recently added waiting request. Some research has shown that LIFO is often better than FIFO request handling.